Rework/simplify unwind infrastructure, implement Windows unwind, and add Windows/new-backend to CI.

cfallin opened the pull request:
Our previous implementation of unwind infrastructure was somewhat
complex and brittle: it parsed generated instructions in order to
reverse-engineer unwind info from prologues. It also relied on some
fragile linkage to communicate instruction-layout information that VCode
was not designed to provide.

A much simpler, more reliable, and easier-to-reason-about approach is to
embed unwind directives as pseudo-instructions in the prologue as we
generate it. That way, we can say what we mean and just emit it
directly.

The usual reasoning that leads to the reverse-engineering approach is
that metadata is hard to keep in sync across optimization passes; but
here, (i) prologues are generated at the very end of the pipeline, and
(ii) if we ever do a post-prologue-gen optimization, we can treat unwind
directives as black boxes with unknown side-effects, just as we do for
some other pseudo-instructions today.

It turns out that it was easier to just build this for both x64 and
aarch64 (since they share a factored-out ABI implementation), and wire
up the platform-specific unwind-info generation for Windows and SystemV.
Now we have simpler unwind on all platforms and we can delete the old
unwind infra as soon as we remove the old backend.

There were a few consequences to supporting Fastcall unwind in
particular that led to a refactor of the common ABI. Windows only
supports naming clobbered-register save locations within 240 bytes of
the frame-pointer register, whatever one chooses that to be (RSP or
RBP). We had previously saved clobbers below the fixed frame (and below
nominal-SP). The 240-byte range has to include the old RBP too, so we're
forced to place clobbers at the top of the frame, just below saved
RBP/RIP. This is fine; we always keep a frame pointer anyway because we
use it to refer to stack args. It does mean that offsets of fixed-frame
slots (spillslots, stackslots) from RBP are no longer known before we do
regalloc, so if we ever want to index these off of RBP rather than
nominal-SP because we add support for `alloca` (dynamic frame growth),
then we'll need a "nominal-BP" mode that is resolved after regalloc and
clobber-save code is generated. I added a comment to this effect in
`abi_impl.rs`.

The above refactor touched both x64 and aarch64 because of shared code.
This had a further effect in that the old aarch64 prologue generation
subtracted from `sp` once to allocate space, then used stores to `[sp, offset]`
to save clobbers. Unfortunately the offset only has 7-bit
range, so if there are enough clobbered registers (and there can be --
aarch64 has 768 bytes of registers; at least one unit test hits this)
the stores/loads will be out-of-range. I really don't want to synthesize
large-offset sequences here; better to go back to the simpler
pre-index/post-index `stp r1, r2, [sp, #-16]` form that works just like
a "push". It's likely not much worse microarchitecturally (dependence
chain on SP, but oh well) and it actually saves an instruction if
there's no other frame to allocate. As a further advantage, it's much
simpler to understand; simpler is usually better.

This PR adds the new backend on Windows to CI as well.

bjorn3 commented:
You are missing debuginfo for the `mov %rsp,%rbp` instruction in the prologue.

Another thing: How hard would implementing asynchronous exceptions with landingpads for cleanup be after this PR compared to before? For example [Theseus OS](https://github.com/theseus-os/Theseus) depends on unwinding after exceptions for fault recovery. LLVM's lack of asynchronous exception support is responsible for a significant part of the failed recovery attempts after a fault is injected. (See table 1 of https://kevinaboos.web.rice.edu/docs/theseus_boos_osdi2020.pdf)

cfallin commented:
> You are missing unwindinfo for the `mov %rsp,%rbp` instruction in the prologue.

Can you be more specific about what you believe is missing?

The unwind directives in the Windows case are: `PushRegister` with `rbp` (this corresponds with the `push rbp`); `SetFPReg` with `rbp` (this corresponds with the `mov rbp, rsp`); `StackAlloc` if a frame is allocated; and a `SaveRegister` for each clobber-save. (Code [here](https://github.com/cfallin/wasmtime/blob/208e8372018724c176cdfae28d83e9d1a956ab47/cranelift/codegen/src/isa/unwind/winx64.rs#L337).) The CFI directives in the SysV case are: `CfaOffset` and `Offset` for the `push rbp`, indicating that the CFA is at RSP+16 and that RBP is saved at `CFA-16`; `Offset` for `LR` on aarch64 with `CFA-8`; `CfaRegister` with `RBP` on the `mov rbp, rsp` indicating that the CFA is now defined in terms of rbp; and `Offset` for each saved clobbered register. (Code [here](https://github.com/cfallin/wasmtime/blob/208e8372018724c176cdfae28d83e9d1a956ab47/cranelift/codegen/src/isa/unwind/systemv.rs#L134).)

In the latter case (SysV), I looked at the output of `gcc -g -S` and `clang -g -S` to match the CFI directives they were generating. On all platforms tests are passing (and a number of unit tests actually exercise unwind, as I can attest from debugging!), so it would seem that whatever directives that are needed are present.

But if there's something non-idiomatic about the output in some case, then I'm happy to fix it!

> Another thing: How hard would implementing asynchronous exceptions with landingpads for cleanup be after this PR compared to before? For example [Theseus OS](https://github.com/theseus-os/Theseus) depends on unwinding after exceptions for fault recovery. LLVM's lack of asynchronous exception support is responsible for a significant part of the failed recovery attempts after a fault is injected. (See table 1 of https://kevinaboos.web.rice.edu/docs/theseus_boos_osdi2020.pdf)

That's interesting. I expect that we'll need to tackle landingpads at some point, driven at least by Wasm exceptions (when that proposal advances) as well as by other use-cases like this one. The complexity around them, though (correctly modeling control-flow, including during optimizations) is something we would need to carefully study, and fully asynchronous exceptions would also require CFI info for epilogues (we don't generate this today because we never unwind in the middle of an epilogue -- there's simply no way for the CLIF to insert any user action there, and there are no asynchronous interrupts). So I wouldn't call it *easy*, exactly, and I can't personally justify spending time on it in the short term, but it should be addressed at some point.

bjorn3 commented:
> Can you be more specific about what you believe is missing?

I only saw the `push rbp` and `sub rsp, ....` instructions have a corresponding unwindinfo entry. In addition if I stepped to the `mov rbp, rsp` there was an extra `???` frame in the backtrace produced by gdb just below the top frame.

> So I wouldn't call it easy, exactly, and I can't personally justify spending time on it in the short term, but it should be addressed at some point.

I understand that it is not a priority right now. I just wanted you to consider if landing this PR would make it harder to implement that to avoid possibly having to revert back to the previous approach.

cfallin commented:
> > Can you be more specific about what you believe is missing?
> 
> I only saw the `push rbp` and `sub rsp, ....` instructions have a corresponding unwindinfo entry. In addition if I stepped to the `mov rbp, rsp` there was an extra `???` frame in the backtrace produced by gdb just below the top frame.

Hmm. Possibly this has to do with the placement of the unwindinfo opcodes; the Push and SetFP are both after the `push rbp / mov rbp, rsp` sequence, whereas if a debugger is stopped in between the two, it should see the Push but not the SetFP. I'll fix that tomorrow and try stepping through it. Thanks for the report!

(I did check with clang on Msys/MinGW just now and the unwind opcodes we're generating are exactly the same as what it generates for its prologue, modulo the above placement issue, so I hope this does it...)

> > So I wouldn't call it easy, exactly, and I can't personally justify spending time on it in the short term, but it should be addressed at some point.
> 
> I understand that it is not a priority right now. I just wanted you to consider if landing this PR would make it harder to implement that to avoid possibly having to revert back to the previous approach.

Ah, I see; if anything it will be easier with a more direct approach, I think. Most of the work for landingpads will be dealing with the control-flow implications in the rest of the compiler; adding the exception handler info to the tables will be relatively straightforward, I think.

cfallin commented on cranelift/codegen/meta/src/shared/settings.rs:
```diff
@@ -235,6 +235,18 @@ pub(crate) fn define() -> SettingGroup {
     );
 
+    settings.add_bool(
+        "no_unwind_info",
```
Done, thanks!

cfallin commented:
@bjorn3, I've updated and verified now that stepping through instruction-by-instruction with `gdb`, the backtrace/unwind is precise at every program point. I went ahead and added epilogue info too since this was not much more work. (Note also that the old impl was missing epilogue info because it was tricky to work out exactly where the epilogues were and plumb this through to the right place.)

I also ripped out the old SysV unwind infra in the new-x64 and aarch64 backends; this was dead code, replaced by this unified new implementation, but I had not removed it previously. Happy to see diff-stats show net-negative lines of code after doing that even with Fastcall and epilogues added!

bjorn3 commented on cranelift/codegen/src/isa/x64/inst/unwind/systemv.rs:
```diff
@@ -82,123 +80,16 @@ pub fn map_reg(reg: Reg) -> Result<Register, RegisterMappingError> {
-            _ => panic!("expected unwind information"),
-        };
-
-        assert_eq!(format!("{:?}", fde), "FrameDescriptionEntry { address: Constant(4321), length: 22, lsda: None, instructions: [(1, CfaOffset(16)), (1, Offset(Register(6), -16)), (4, CfaRegister(Register(6))), (15, RememberState), (17, RestoreState)] }");
```
Could you restore these tests for the actual unwind info being emitted?

cfallin replied:
Sorry, not sure why I removed that along with the dead code. Added back.

cfallin commented:
Updated again; I realized that we need a few more ops around mid-function epilogues (`RememberState` / `RestoreState`) because the CFA parser doesn't trace control flow, but rather just scans the function linearly.

cfallin commented:
So it turns out that the epilogue CFA info was causing segfaults on CI but not locally, and I wasn't able to find the root cause after much head-scratching. Since epilogue info is not necessary for our unwind needs (it only becomes necessary if one has a trap somewhere in the middle of the `mov rsp, rbp / pop rbp / ret` sequence, or if one wants to instruction-step through that sequence and have gdb print a nice backtrace while in the middle), and since the current state of both the new x64 backend and the aarch64 backend is that no epilogue CFA info is provided, I've gone ahead and re-simplified to provide only prologue info.

alexcrichton commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -535,21 +550,39 @@ impl ABIMachineSpec for AArch64MachineDeps {
+        // because we cannot guarantee that the offset of a clobber-save slot
+        // will be within a SImm7 (+127/-128 byte) offset range of sp after
+        // allocating the whole frame. Because clobbers are at the top of the
+        // frame (due to unwind-info restrictions), this more general scheme is
```
Technically this restriction is just on Windows x86_64, right? (even if we add AArch64 support for Windows that may have not have the same retrictions?)

cfallin replied:
Yes, that's true, but it would be significantly more complex to make the factored-out ABI implementation conditional on platform and do a different thing on aarch64. (It's a tradeoff for sure -- we could fork aarch64's code, as we used to have, but then that doubles the maintenance and update cost in other cases.)

alexcrichton replied:
Nah this is definitely fine to have, just wanted to double-check my understanding!

cfallin replied:
Updated comment.

alexcrichton commented on cranelift/codegen/src/isa/aarch64/abi.rs:
```diff
@@ -535,21 +550,39 @@ impl ABIMachineSpec for AArch64MachineDeps {
+        // We use pre-indexed addressing modes here, rather than the possibly
+        // more efficient "subtract sp once then used fixed offsets" scheme,
+        // because we cannot guarantee that the offset of a clobber-save slot
+        // will be within a SImm7 (+127/-128 byte) offset range of sp after
```
This mentions `SImm7` but below it uses `SImm7Scaled` which seems to have a range of (63 * 8) to -(64 * 8). Is that a bug in the documentation or am I missing the usage of `SImm7`? More-or-less I'm wondering if given a 504 byte range we could do everything as a positive offset relative to `sp`?

cfallin replied:
Yes, my mistake, it is actually a scaled offset; I'll update the comment, thanks!

That said, going back to an sp-offset scheme requires two assumptions: (i) the subset of registers that are callee-save fits within 512 bytes; this is true for SysV but may be a subtle footgun in the future; and (ii) that we are alright with a two-stage thing, where we decrement SP once and save clobbers, then decrement again for the frame, unless we special-case the "everything is in range" case and collapse those two into one. I'd rather not have that complexity if we can avoid it!

Basically my high-level bit here is: try to find a common set of simple lowerings and shared layouts/schemes that does not have a lot of special cases for different ABIs, different architectures, different in-range/out-of-range cases; bugs and unmaintainable complexity lie that way. The optimized all-in-one `sub sp` stuff that this PR removes was already very hard for me to re-internalize when I had to change it for this work. So I'd rather standardize on a simple "push the things one at a time" lowering, and have the same frame layout everywhere, to keeps us all sane :-)

alexcrichton replied:
Nah that makes sense, and I think it's fine to keep as-is. I definitely agree having a known-correct thing is the highest priority. If it turns out that we want to change this in the future for perf reasons or something like that we can cross that bridge when we get there.

cfallin replied:
Updated comment here -- yes, you're right, the range is actually 504 bytes. I've written out the reasoning here w.r.t. simplicity and avoiding a two-stage frame setup with big frames.

alexcrichton commented on cranelift/codegen/src/isa/unwind.rs:
```diff
@@ -86,3 +89,89 @@ pub mod input {
+    /// A return has just occurred. This can be used by unwind
+    /// formats (like System V) to know where to delimit unwind-state
+    /// transitions around mid-function returns.
+    Ret,
```
Since this and `PopOldFP` aren't actually used anywhere today, should they be removed for now?

I'd imagine that unwind info for the epilogue would also need to record that the saved registers no longer need to restored, but I don't think there's any emission for that just yet.

cfallin replied:
Yep, this is left over from my intermediate PR state where I tried to handle epilogues too. Happy to remove!

cfallin replied:
Yep, I've removed all traces of the epilogue attempt now.

alexcrichton commented on cranelift/codegen/src/isa/unwind.rs:
```diff
@@ -86,3 +89,89 @@ pub mod input {
+    /// positive offsets from the base of the frame are allowed for
+    /// `SaveReg`, but the FP register itself can point to any part
+    /// of the frame.
+    CreateFPFrame {
```
Ok so after reading the various ABI implementations I think I understand a little better, but not entirely. It seems like this is primarily around for Windows and calculating adjustments on Unix, right?

For the Unix case (which I think I understand better) this seems like it should be something like `DeclareCobberFrameSize { size: u8 }` where it's basically just declaring that all saves to come later are relative offsets within a frame of size `size`, allocated at the top of the stack just after the frame pointer was pushed.

For Windows I'm much more unfamiliar with the unwinding stuff. I'm a bit confused, though, on why everything is `u8` here. It seems to be because of Windows but the [exception codes](https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=msvc-160) all have short/far variants which seems like it's trying to account for offsets which are larger than `u8` perhaps, but I may be misunderstanding something. For Windows though it seems like this should be two instructions, where one signifies the `mov %rsp, %rbp` happened and the other happens after the stack decrement.

I'll admit though I'm pretty confused on Windows, and I probably just need to sit down and read through their documentation some more. I'm not sure if we're doing RSP-relative addressing or whatever the second mode is mentioned [at the top of this](https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64?view=msvc-160#unwind-operation-code) where something is FP relative. I always seem to get lost in the details thinking about stuff like this.

cfallin replied:
So WIndows forces our hand in a few ways here:

* The way that clobbers are located is as a *positive* offset from the "frame", wherever we choose to define the frame. Since we place `RBP`/`FP` just below the return address so that we have known offset addressing to stack args, and we place clobbers below `RBP`, we have to either define the frame in terms of `RSP` or as some offset below `RBP`. The former is difficult for several reasons but basically our nominal-SP tracking would not interact well with unwind info due to the order we compute and finalize things. So we define the frame in terms of `RBP`, but `RBP` is at the *top* of it.
* This is what `fp_offset` means (it comes directly from the "frame register offset" field in `UNWIND_INFO`, which is also the source of the `u8`-sized limit): when we declare the frame, its base is at `RBP - fp_offset`. Equivalently, this is saying that the frame pointer is at the "frame pointer offset" into the frame.
* Note that on SysV platforms, the frame is defined by the "CFA" (call-frame address) which is spec-defined to be the SP value at entry, so instead `fp_offset` is used to adjust the offsets for clobber-saves.
* The `fp_offset` size limit (actually 240 bytes) is the reason that clobbers had to be moved to just below `RBP` as well; otherwise we could not define a frame far enough away from `RBP` to straddle our other fixed-frame slots.

cfallin replied:
Regarding `fp_offset` in particular, I'll go over the doc comments some more and see if I can add detail. Perhaps what we really need is just a small tutorial on unwind formats on Windows and SysV and how frames are computed in both cases :-)

alexcrichton replied:
Ok thanks for the info! This helps dispel some of my confusion. I'm familiar with the CFA in dwarf but the "cfa equivalent" was actually in the other direction on Windows and I was getting confused how these two things map to each other. I think for now some extra documentation for how the field is expected to be interpreted on Windows and Unix should be fine myself.

But for the `u8`-ness, to make sure I understand, we're declaring to Windows "all register saves, which occur at positive offsets, are relative to XX" where XX is `RBP-thing` and `thing` is limited to 8 bits? We then have an *additional* requirement later on that when we save a register it can be saved at most 240 bytes away? (I couldn't actually find where that second restriction arose myself)

cfallin replied:
Re: offset restrictions, only the former (`RBP - offset`); once we establish the frame we can declare clobber-saves at any 32-bit offset. But combining this with the fact that we place `RBP` *above* clobber-saves gives us the limit on clobber offsets.

In hindsight, perhaps it's better to have explicit asserts that all offsets are in-range for the unwind info we're generating, and give these offset fields a `u32` or so; that would document these limits more explicitly I think!

cfallin replied:
(Err, "any positive 32-bit offset" above.)

peterhuene replied:
Alternatively we could stop establishing a frame pointer until such a time we actually need one on Windows (the save offsets encoded in the unwind information would then be positive offsets from RSP rather than FP-[0..240], thereby eliminating the "saves have to be near enough to the frame pointer establishment" problem).  That's a whole can of worms, though, I imagine.

peterhuene replied:
In fact, the old unwind info emitted for Windows explicitly didn't encode establishing a frame pointer (despite the prologue doing so) to side step the 240 offset limitation, thereby encoding all of the register saves as positive offsets from the stack pointer.  This would obviously break when the frame pointer is required, such as dynamically-sized frames.

alexcrichton replied:
Aha ok that makes a lot of sense now, where the 240 comes from. I do think there's definitely room for improvement here, but there's no need to happen in this PR I think. I'd be happy with some more comments myself, and then seems reasonable to open tracking issues if necessary.

cfallin replied:
So I've revamped the doc-comment for this enum and added the "offset upward / offset downward" dual fields for the CreateNewFrame pseudo-inst. Hopefully this is a bit more clear now!

Agreed that we could eventually look into `-fomit-frame-pointer` codegen, though that's a deeper rabbithole in the ABI code (as is `alloca` support); one bridge at a time :-)

alexcrichton reviewed (commented):
Sure would be happy to!

FWIW while I'm pretty confident in the dwarf unwind information I'm much less confident in the Windows bits. This was the first time I looked at the documentation for windows unwinding codes and to really get a good understanding I'd need to set aside a bigger chunk of time to dig in and see how what we do maps to the specification and such. I suspect it's all fine since tests are all green, but there were definitely lots of places I couldn't connect the dots to reason about why things were the way they were.

cfallin commented on cranelift/codegen/src/isa/unwind/systemv.rs:
```diff
@@ -112,7 +122,62 @@ pub struct UnwindInfo {
+                }
+            }
+            &UnwindInst::CreateFPFrame { fp_offset } => {
+                // Define CFA as FP+16
```
Yes, I'm thinking now perhaps we just give two offsets to describe FP: its offset above the start of clobbers, and its offset below SP at entry. These are the two values needed to produce either Windows or SysV info so it should make everything else completely platform-agnostic.

cfallin replied:
Updated as noted -- two offsets given now.

cfallin commented on cranelift/codegen/src/isa/unwind/winx64.rs:
```diff
@@ -285,6 +324,71 @@ impl UnwindInfo {
+                frame_register_offset = fp_offset;
+                unwind_codes.push(UnwindCode::SetFPReg { offset });
+                if fp_offset > 0 {
+                    unwind_codes.push(UnwindCode::StackAlloc {
```
Perhaps it should just be called `CreateFrame` -- it was meant to be a single macro-op that captured anything a specific unwind format needs for the prologue. The close mapping between `sub %rsp` and `StackAlloc` was the sort of thing I was hoping to avoid in this simplification :-) Given that we pin the frame to `RBP` with the `SetFPReg`, the `StackAlloc` serves only to let the unwinder know that more stack space is being used; so as long as it comes after the `SetFPReg` and before any uses of the clobber space, it shouldn't matter exactly where it occurs, I think.

alexcrichton replied:
Ok I think that's fine in the meantime anyway. I don't think it's required we get this 100% right but it might just be worth in the documentation to mention that it's sort of a macro-op for now.

cfallin replied:
I did one better and just removed the `StackAlloc`; it doesn't seem to actually be needed once we pin the frame to `RBP`.

cfallin commented on cranelift/codegen/src/isa/x64/abi.rs:
```diff
@@ -474,22 +496,29 @@ impl ABIMachineSpec for X64ABIMachineSpec {
+            });
+        }
+
+        // Adjust the stack pointer downward for clobbers.
```
Yep, that's a stale comment, thanks!

cfallin replied:
Fixed!

cfallin commented on cranelift/filetests/filetests/isa/aarch64/floating-point.clif:
```diff
@@ -1,4 +1,5 @@
 test compile
+set unwind_info=false
```
Done; changed just for filetests (I think it's too surprising to change the default overall given that it's needed for backtraces and correct operation generally).

cfallin commented on cranelift/codegen/src/isa/unwind/winx64.rs:
```diff
@@ -1,6 +1,10 @@
+#![allow(dead_code)]
```
Ah, yes, I was trying to get rid of some unused-code warnings in the new-backend PR and this percolated backward during some rebasing magic. Removed!

cfallin commented on cranelift/codegen/src/isa/unwind/winx64.rs:
```diff
@@ -143,9 +179,9 @@ pub(crate) enum MappedRegister {
 /// Maps UnwindInfo register to Windows x64 unwind data.
-pub(crate) trait RegisterMapper {
+pub(crate) trait RegisterMapper<Reg> {
     /// Maps RegUnit.
```
Fixed.

cfallin commented on cranelift/codegen/src/isa/unwind/systemv.rs:
```diff
@@ -1,6 +1,10 @@
+#![allow(dead_code)]
```
Removed, thanks!

cfallin commented on cranelift/codegen/src/isa/unwind.rs:
```diff
@@ -86,3 +89,89 @@ pub mod input {
+/// push rbp
+/// unwind PushOldFP sp_offset_to_entry=8
+/// mov rbp, rsp
+/// unwind CreateFPFrame (BP=sp; [BP] = oldBP)
```
Updated along with the doc and field-name clarity changes noted below.

cfallin commented on cranelift/codegen/src/isa/unwind/winx64.rs:
```diff
@@ -285,6 +324,71 @@ impl UnwindInfo {
+                if fp_offset > 0 {
+                    unwind_codes.push(UnwindCode::StackAlloc {
+                        offset,
+                        size: fp_offset as u32,
```
Done!

cfallin commented on cranelift/codegen/src/isa/unwind/winx64.rs:
```diff
@@ -51,6 +55,11 @@ pub(crate) enum UnwindCode {
         offset: u8,
```
I've changed this to `instruction_offset` everywhere now. The tuple refactor is a bit more intrusive but I can do that instead if you'd prefer!

cfallin commented on cranelift/codegen/src/isa/x64/abi.rs:
```diff
@@ -433,26 +433,48 @@ impl ABIMachineSpec for X64ABIMachineSpec {
+        if flags.unwind_info() {
+            insts.push(Inst::Unwind {
+                inst: UnwindInst::PushOldFP {
+                    sp_offset_to_caller: 16,
```
Done.

cfallin commented on cranelift/codegen/src/isa/x64/abi.rs:
```diff
@@ -498,18 +527,21 @@ impl ABIMachineSpec for X64ABIMachineSpec {
         for reg in &clobbered {
             let r_reg = reg.to_reg();
-            match r_reg.get_class() {
+            let off = match r_reg.get_class() {
```
Ah, yeah, that's a silly bit of verbosity; fixed.

cfallin commented on cranelift/codegen/src/isa/unwind/systemv.rs:
```diff
@@ -112,7 +122,62 @@ pub struct UnwindInfo {
+                    offset,
+                    CallFrameInstruction::Offset(mr.fp(), -(sp_offset_to_caller as i32)),
+                ));
+                if let Some(lr) = mr.lr() {
```
Added to the doc-comment! I've also removed this to `PushFrameRegs` rather than `PushOldFP`.

cfallin commented on cranelift/codegen/src/isa/unwind/systemv.rs:
```diff
@@ -112,7 +122,62 @@ pub struct UnwindInfo {
+                let off = frame_base + (frame_offset as i32);
+                instructions.push((offset, CallFrameInstruction::Offset(reg, off)));
+            }
+            _ => {}
```
Updated.

cfallin commented on cranelift/codegen/src/isa/unwind.rs:
```diff
@@ -86,3 +89,89 @@ pub mod input {
+/// unwind CreateFPFrame (BP=sp; [BP] = oldBP)
+/// sub rsp, 32
+/// mov [rsp+0], r12
+/// unwind SaveReg [BP-32], r12
```
Updated.

cfallin commented:
Updated based on comments; thanks!

peterhuene commented on cranelift/filetests/filetests/isa/x64/fastcall.clif:
```diff
@@ -91,10 +104,12 @@ block0(v0: i64, v1: i64, v2: i64, v3: i64, v4: i64, v5: i64):
 ; nextln: subq    $$16, %rsp
 ; nextln: movq    %rsi, 0(%rsp)
-; nextln: virtual_sp_offset_adjust 16
+; nextln: unwind SaveReg { clobber_offset: 0
```
Is this match (and others like it) intentionally truncated?

peterhuene commented on cranelift/codegen/src/machinst/abi_impl.rs:
```diff
@@ -22,26 +22,41 @@
+//! without a long-form sequence to synthesize an arbitrary
+//! offset). We also need clobber-save registers to be "near" the
+//! frame pointer: Windows unwind information requires it to be within
+//! 256 bytes of RBP. Finally, it is not allowed to access memory
```
```suggestion
//! 240 bytes of RBP. Finally, it is not allowed to access memory
```

cfallin replied:
Done.

alexcrichton reviewed (approved):
Thanks so much for all the new comments! Definitely makes sense to me now :)

peterhuene reviewed (approved):
This looks fantastic!

I glanced over the aarch64-related changes for the most part, although I did pay attention to the common unwind representation and read over your detailed comments explaining things.

Therefore I paid the most attention to the x64 backend changes and the changes to the unwind generation for both SystemV and Windows and have just three minor comments.

cfallin commented on cranelift/codegen/src/isa/unwind.rs:
```diff
@@ -86,3 +89,146 @@ pub mod input {
+/// info for epilogues is only necessary if one expects to unwind while within
+/// the last few instructions of the function (after FP has been restored) or
+/// if one wishes to instruction-step through the epilogue and see a backtrace
+/// at every point. This is not necessary for correct operation otherwise and so
```
Good point, done!

cfallin commented:
@sunfishcode it looks like there might be an intermittent test with `wasi_cap_std_sync` on Windows, namely `poll_oneoff_stdio` -- see [logs](https://github.com/bytecodealliance/wasmtime/runs/2091391143?check_suite_focus=true#step:10:1009). Watching the latest run to see if it fares better this time...

uweigand commented:
Hi @cfallin, I finally got around to updating my s390x patch to this new unwind method.  This ran into a couple of problems, so I'm wondering whether I'm on the right track here.  I'd appreciate your input ...

1. Frame size definition.    The `total_frame_size` value returned form `frame_size` is documented to equal the "distance between FP and nominal SP", and users (even outside of my new back-end) appear to rely on that fact, e.g. it is ultimately used to implement the `get_nominal_sp_to_fp` routine used in `spillslots_to_stack_map`.   However, with the new approach, the stack space used for clobbers (`clobber_size`) now lies between the nominal SP and the FP, so I believe this value now needs to be added to `total_frame_size`.     (At least in the non-baldrdash case.  I'm not really familiar with how baldrdash works, in particular in the new approach ...)
2. Accumulate outgoing args.  We've added this feature a while back since I'm using it in my back-end.  This patch completely removed the feature again, breaking my back-end.  Would you accept a patch adding it back in the new approach?
3. Operation without frame pointer.   I don't really need or want a frame pointer, but in the new approach I'm basically forced to define a FP register.   Would it be OK to add support for non-FP operation?   One option would be to make the FP register optional (like the LR currently is) and update the handling of `UnwindInst::PushFrameRegs` and `UnwindInst::DefineNewFrame` if no FP is present -- this is what I've implemented so far.  Another option would be to define a new `UnwindInst::AllocStack` that I could use instead of those two if there is no FP.
4. Negative offsets to `UnwindInst::SaveReg`.    In particular in the absence of a FP (where I'm not using `DefineNewFrame`), the offsets to `SaveReg` are more naturally expressed relative to the incoming SP (or CFA).  But then they will be negative, which makes it a bit awkward to use the current interface which defines the offsets as `u32`.

cfallin commented:
Hi @uweigand -- thanks for the efforts to update and sorry about the churn!

I'll note first as an aside that I'm quite open to merging the s390x backend sooner rather than later, if you think it is at a place for that, as it would allow me and others to consider its constraints and update it along with the others when refactors occur. I am of course also happy to work out followup changes to resolve issues after-the-fact, as we're doing here, but e.g. if I had known that the "accumulate outgoing args" mechanism was used by s390x as well, it might have affected my cleanup/simplification calculus a bit (sorry about that!).

In particular, I'm currently working on an update to the register allocator, which involves some churn in the way it is used in the lowering backends. I'm happy to do the update in all backends in-tree, including s390x if it is merged; keeping it out-of-tree creates more rebasing pain for you as we refactor and evolve, which I do regret :-/

To the specific questions:

> Frame size definition ... [add `clobber_size` to nominal-SP-to-FP distance]

Yes, I think so, though without seeing your ABI implementation in more detail it's hard to say for sure. Are you using `abi_impl.rs` or implementing the `ABICaller` / `ABICallee` traits directly?

> Accumulate outgoing args

Sorry about this! I worked to simplify the common ABI implementation as I refactored it and this seemed to me to be implementing something we could do in a simpler way, at least on the x64 and aarch64 backends. I hadn't realized it was used on s390x. I'm happy to take a patch to add it back, or, if it is simpler, perhaps the s390x backend could directly implement this in its ABI code (without indirecting through the traits)?

> Operation without frame pointer

We do definitely want to make this possible eventually. Without thinking through the implications a bit more, I don't have a strong opinion which of the two approaches you name makes more sense, but my initial thought is that perhaps we add `AllocStack` as you suggest and then omit `PushFrameRegs`. The SysV unwind info generator then just needs to track that it is in FP-less mode and continue to define the CFA in terms of SP instead.

> Negative offsets to `UnwindInst::SaveReg`

To unify the view between the different platforms, I think it does make sense to continue to view the clobber locations in terms of a "clobber frame" that starts below them; I think that if we make `DefineNewFrame` work in an FP-less environment (but still have the same notion of an offset from some anchor to the start of clobbers) this should be possible? Or is there a fundamental conflict here?

(This is useful pathfinding for our eventual omit-FP optimization on other architectures too, by the way, so thank you for probing these issues now!)

uweigand commented:
> Hi @uweigand -- thanks for the efforts to update and sorry about the churn!

No problem, thanks for the reply!
 
> I'll note first as an aside that I'm quite open to merging the s390x backend sooner rather than later, if you think it is at a place for that, as it would allow me and others to consider its constraints and update it along with the others when refactors occur. I am of course also happy to work out followup changes to resolve issues after-the-fact, as we're doing here, but e.g. if I had known that the "accumulate outgoing args" mechanism was used by s390x as well, it might have affected my cleanup/simplification calculus a bit (sorry about that!).
> 
> In particular, I'm currently working on an update to the register allocator, which involves some churn in the way it is used in the lowering backends. I'm happy to do the update in all backends in-tree, including s390x if it is merged; keeping it out-of-tree creates more rebasing pain for you as we refactor and evolve, which I do regret :-/

I agree, in particular because there now also seems to be renewed interest in s390x support in wasmtime due to additional use cases.  I'll try to get things ready for an initial merge soon, hopefully within the next few weeks.

> To the specific questions:
> 
> > Frame size definition ... [add `clobber_size` to nominal-SP-to-FP distance]
> 
> Yes, I think so, though without seeing your ABI implementation in more detail it's hard to say for sure. Are you using `abi_impl.rs` or implementing the `ABICaller` / `ABICallee` traits directly?

I'm using `abi_impl.rs` just like x64 and aarch64 do.  I had thought that the problem with the get_nominal_sp_to_fp routine should also occur on those platforms, but at a second glance, it probably does not: the only user of this routine in common code is 
spillslots_to_stack_map, but this uses the value only as an upper bound of the stack map size.  Given that there are no stack slots in the clobber area, for this particular use case the difference probably does not matter.

On my platform, I have another use case in target-specific code.  This again has to do with the fact that I am not using a frame pointer register, and therefore the code generator has to rewrite accesses to the incoming argument slots (which are expressed by common code via offsets relative to the FP) in terms of offsets relative to the (nominal) SP.  This is done by simply adding the nominal_sp_to_fp offset, which used to work fine, but has now stopped working.

So we can either fix the frame_size implementation so that nominal_sp_to_fp again returns the actual offset between those two points in the stack, or else I'll have to find some other way to pass the correct offset including clobbers to the back-end codegen (e.g. via a pseudo instruction).

In case you agree with changing frame_size, here's a PR to do so: https://github.com/bytecodealliance/wasmtime/pull/2836

> > Accumulate outgoing args
> 
> Sorry about this! I worked to simplify the common ABI implementation as I refactored it and this seemed to me to be implementing something we could do in a simpler way, at least on the x64 and aarch64 backends. I hadn't realized it was used on s390x. I'm happy to take a patch to add it back, or, if it is simpler, perhaps the s390x backend could directly implement this in its ABI code (without indirecting through the traits)?

Here's a PR to add this support back: https://github.com/bytecodealliance/wasmtime/pull/2837

I'd be happy to implement this solely in the back-end, but the main problem I was having with this is that there does not appear to be any back-end specific (mutable) **state** associated with an ABICallee.   I need this state to track the required outgoing argument space as we go through the whole function and lower it; each call instruction that is found will have to update the required size accordingly.  Finally, in the prologue and epilogue code generator I need to access that accumulated value so I can alloc/dealloc the correct amount of extra stack space.

Without the common code implementation, do you have a suggestion where I should hold this extra state?
 
> > Operation without frame pointer
> 
> We do definitely want to make this possible eventually. Without thinking through the implications a bit more, I don't have a strong opinion which of the two approaches you name makes more sense, but my initial thought is that perhaps we add `AllocStack` as you suggest and then omit `PushFrameRegs`. The SysV unwind info generator then just needs to track that it is in FP-less mode and continue to define the CFA in terms of SP instead.
> 
> > Negative offsets to `UnwindInst::SaveReg`
> 
> To unify the view between the different platforms, I think it does make sense to continue to view the clobber locations in terms of a "clobber frame" that starts below them; I think that if we make `DefineNewFrame` work in an FP-less environment (but still have the same notion of an offset from some anchor to the start of clobbers) this should be possible? Or is there a fundamental conflict here?

Here's a PR to add support for operation without frame pointer along the lines suggested above: https://github.com/bytecodealliance/wasmtime/pull/2838

In particular, this extends DefineNewFrame as you suggest for FP-less operation, and adds a new StackAlloc operation.  With those two, I can define the unwind information in my backend in a pretty straightforward manner.

cfallin commented:
The PRs all look good -- will land all three. Thanks!

uweigand commented:
Thank you!
